ksqldb-engine/src/main/java/io/confluent/ksql/util/SharedKafkaStreamsRuntimeImpl.java

```diff
@@ -94,20 +99,54 @@ public void register(
               Throwables.getStackTraceAsString(e),
               errorType
           );
-      for (BinPackedPersistentQueryMetadataImpl query: collocatedQueries.values()) {
-        query.getListener().onError(collocatedQueries.get(query.getQueryId()), queryError);
+
+      final String queryId;
+      if (e instanceof StreamsException && ((StreamsException) e).taskId().isPresent()) {
+        queryId = ((StreamsException) e).taskId().get().topologyName();
+      } else {
+        queryId = null;
+      }
+
+      if (queryId == null) {
+        for (BinPackedPersistentQueryMetadataImpl query : collocatedQueries.values()) {
+          query.getListener().onError(collocatedQueries.get(query.getQueryId()), queryError);
```

Comment:
why not just use the query's `setQueryError` ? Then the `getQueryErrors` is much simpler because you already have all the relevant errors in the queue

Reply:
You do mean just remove the `runtimeErrors` queue altogether and add any runtime-wide errors to all query error queues, yes? SGTM, the errors should be pretty small so not worth trying to de-duplicate